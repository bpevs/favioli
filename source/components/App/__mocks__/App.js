export const App = () => <div></div>
