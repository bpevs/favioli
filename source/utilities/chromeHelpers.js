const { storage, tabs } = chrome;

const defaultSettings = {
  overrides: [],
  overrideAll: false,
};

export function getTab(tabId) {
  return new Promise(resolve => tabs.get(tabId, resolve));
}

export function getSettings(settings) {
  const toGet = Object.assign({}, defaultSettings, settings);
  return new Promise(resolve => storage.sync.get(toGet, resolve));
}

export function setSettings(toSet) {
  return new Promise(resolve => storage.sync.set(toSet, resolve));
}
