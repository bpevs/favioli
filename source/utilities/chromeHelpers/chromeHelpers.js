const { storage, runtime, tabs } = (typeof chrome ? chrome : browser);
import getEmojiFromLegacyString from "../../constants/emoji2Name"


const defaultOptions = {
  flagReplaced: false,
  overrideAll: false,
  overrides: [],
};


/**
 * Get information about a tab
 * @param {number} tabId
 */
export function getTab(tabId) {
  return new Promise(resolve => tabs.get(tabId, resolve));
}


/**
 * Get options.
 */
export function getOptions() {
  return new Promise((resolve, reject) => storage.sync.get(
    Object.keys(defaultOptions),
    options => {
      if(runtime.lastError) return reject(runtime.lastError)

      // Legacy Favioli used straight emoji strings. Using the format of
      // EmojiMart allows us great flexibility for future expansion
      const overrides = (options && options.overrides || [])
        .map(override => Object.assign({}, override, {
          emoji: typeof override.emoji === "string"
            ? getEmojiFromLegacyString(options.overrides[0].emoji)
            : override.emoji
        }))

      resolve(Object.assign({}, defaultOptions, options, { overrides }))
    }
  ));
}

/**
 * Set Options
 * @param {object} toSet
 */
export function setOptions(toSet) {
  if (!toSet) return;

  const options = {
    flagReplaced: Boolean(toSet.flagReplaced),
    overrideAll: Boolean(toSet.overrideAll),
    overrides: toSet.overrides || defaultOptions,
  }

  return new Promise((resolve, reject) => storage.sync.set(
    options,
    () => runtime.lastError ? reject(runtime.lastError) : resolve()
  ))
  .then(() => runtime.sendMessage("updated:options"))
}
