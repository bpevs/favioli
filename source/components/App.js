import React from "react"
import { getOptions, setOptions } from "../utilities/chromeHelpers/chromeHelpers";
import Checkbox from "./Checkbox"
import NavBar from "./NavBar"
import OverridesList from "./OverridesList";
const { runtime } = (typeof chrome ? chrome : browser);
import { emojiIndex } from "emoji-mart"
import getEmojiFromLegacyString from "../constants/emoji2Name"

window.allEmojis = {};
window.emojis = {}

export class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      route: window.location.hash.substr(1) || "overrides",
    }
  }

  componentDidMount() {
    // const options = await restoreOptions()
    // this.setState(options)
  }

  _changeRoute(evt) {
    const route = evt.target.href.match(/\#(.*)/)[1]
    this.setState({ route })
  }

  _saveOptions(overrides) {
    console.log(overrides)
    if (overrides && overrides[0] && overrides[0].emoji) {
      console.log(getEmojiFromLegacyString(overrides[0].emoji.native))
    }
  }

  render() {
    const pages = {
      overrides: <OverridesList onChange={this._saveOptions} />,
      settings: (<div>
        <Checkbox />
      </div>),
    }

    return (
      <div>
        <NavBar onSelect={this._changeRoute.bind(this)} />
        <div className="page">
          <h1 className="title">
            {this.state.route.toUpperCase()}
          </h1>
          {pages[this.state.route]}
        </div>
      </div>
    )
  }
}
