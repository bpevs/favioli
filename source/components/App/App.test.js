import { shallow } from "enzyme"
import * as React from "react"
import { App } from "./App"
import { isBrowser } from "../../utilities/chromeHelpers/chromeHelpers"
const { storage } = (isBrowser("CHROME") ? chrome : browser)


beforeEach(() => {
  jest.clearAllMocks()
})

afterEach(() => {
  jest.useRealTimers()
})

test("Should Render App", async () => {
  const wrapper = await shallow(<App />)
  expect(wrapper).toMatchSnapshot()
});

test("Should get options on mount", async () => {
  await shallow(<App />)
  const getOptions = storage.sync.get.mock
  expect(getOptions.calls.length).toBe(1)
})

test("Should Change Route", async () => {
  const wrapper = await shallow(<App />)
  wrapper.setState({ route: "SETTINGS" })
  expect(wrapper).toMatchSnapshot()
})

test("Should save options", async () => {
  const setOptions = storage.sync.get.mock
  const wrapper = await shallow(<App />)
  wrapper.find("button.save").simulate("click")
  expect(setOptions.calls.length).toBe(1)
})

test("Should notify user on save for a short time", async () => {
  jest.useFakeTimers()
  const el = await shallow(<App />)

  expect(el.state("status")).toBe("")
  await el.find("button.save").simulate("click")

  await el.update()
  expect(el.state("status")).toBe("Successfully saved.")
  jest.runAllTimers()

  expect(el.state("status")).toBe("")
})
