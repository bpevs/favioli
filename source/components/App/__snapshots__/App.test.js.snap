// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Change Route 1`] = `
<div>
  <NavBar
    onSelect={[Function]}
    route="SETTINGS"
  />
  <div
    className="page"
  >
    <h1
      className="title"
    >
      SETTINGS
    </h1>
    <div>
      <Checkbox
        checked={true}
        name="flagReplaced"
        onChange={[Function]}
      />
    </div>
    <button
      className="save"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
  <div
    id="status"
  />
</div>
`;

exports[`Should Render App 1`] = `
<div>
  <NavBar
    onSelect={[Function]}
    route="OVERRIDES"
  />
  <div
    className="page"
  >
    <h1
      className="title"
    >
      OVERRIDES
    </h1>
    <OverridesList
      onChange={[Function]}
      overrides={
        Array [
          Object {
            "emoji": Object {
              "colons": ":grinning:",
              "emoticons": Array [],
              "id": "grinning",
              "name": "Grinning Face",
              "native": "ðŸ˜€",
              "skin": null,
              "unified": "1f600",
            },
            "filter": "bookface",
          },
        ]
      }
    />
    <button
      className="save"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
  <div
    id="status"
  />
</div>
`;
